File: Java/EggDroppingPuzzle.java
/*
 * Suppose you have N eggs and you want to determine from which floor in a K-floor building you 
 * can drop an egg such that it doesn't break. You have to determine the minimum number of attempts 
 * you need in order to find the critical floor in the worst case while using the best strategy.
 * There are few rules given below. 
	* An egg that survives a fall can be used again.
	* A broken egg must be discarded.
	* The effect of a fall is the same for all eggs.
	* If the egg doesn't break at a certain floor, it will not break at any floor below.
    * If the eggs breaks at a certain floor, it will break at any floor above.
 *
 * Time complexity of this function: O(N*K^2)
 * Space Complexity of this function: O(N*K) 
 */

package Hacktoberfest2020_.Java;

public class EggDroppingPuzzle {
	public static class Puzzle {

        // Function to calculate the minimum number of attemps
		public static int eggDrop(int n, int k) {
			int eggFloor[][] = new int[n+1][k+1];
			int res;
			
			for(int i = 0; i <= n; i++) {
				eggFloor[i][0] = 0;
				eggFloor[i][1] = 1;
			}
			
			for(int j = 1; j <= k; j++) {
				eggFloor[1][j] = j;
			}
			
			for(int i = 2; i <= n; i++) {
				for(int j = 2; j <= k; j++) {
					eggFloor[i][j] = Integer.MAX_VALUE;
					for(int x = 1; x <= j; x++) {
						res = 1 + Math.max(eggFloor[i-1][x-1], eggFloor[i][j-x]);
						
						if(res < eggFloor[i][j]) {
							eggFloor[i][j] = res;
						}
					}
				}
			}
			
			return eggFloor[n][k];
		}
	}
	
	public static void main(String[] args) {
		int eggs = 2;
		int floors = 10;
		
		System.out.println("Minimum number of trials required: " + Puzzle.eggDrop(eggs, floors));
	}
}

File: C/EggDroppingPuzzle.c
/* Saurabh Vijay
   Github :- https://github.com/SaurabhVj
   Solution of Egg Dropping Puzzle
*/

#include<stdio.h>
#define INT_MAX 1e9+9 

int maxi(int a , int b)
{
   if(a >= b)
    return a ;

   return b ;
}

int solve(int eggs, int floors) 
{ 

    if (floors == 1 || floors == 0) 
        return floors; 
  
  
    if (eggs == 1) 
        return floors; 
  

    int ans = INT_MAX , possible_ans ;
    int i ;
    

    for(i = 1 ; i <= floors ; i++)
    {
      possible_ans = maxi(solve(eggs-1 , i-1) , solve(eggs , floors - i)) ;
      if(possible_ans < ans)
        ans = possible_ans ;
    }
  
    return ans + 1 ;
} 
int main() 
{ 
   int eggs,floors;
   printf("Enter the number of eggs\n") ;
   scanf("%d",&eggs);
   printf("Enter the number of floors\n");
   scanf("%d",&floors);
  

   printf("Minimum attempts required having %d eggs and %d floors are %d\n",eggs,floors,solve(eggs,floors));
    
    return 0; 
} 
